### PR TITLE
Feature: Midi Follow Mode with Midi Feedback

### PR DESCRIPTION
**Pull request by seangoodvibes**
# MIDI Follow Mode with MIDI Feedback

image

## Description:

Master MIDI follow mode whereby after setting a master MIDI follow channel for Synth/MIDI/CV clips, Kit clips, and for Parameters, all MIDI (notes + cc’s) received will be directed to control the active view (e.g. arranger view, song view, audio clip view, instrument clip view). 

- Note: although there are three types of MIDI follow channel's, all three channels will control the instrument of the active context. The three follow channel's allows you to learn different devices to MIDI follow, should you require device specific channel settings.

Comes with a MIDI feedback mode to send updated parameter values on the MIDI follow channel for mapped MIDI cc's. Feedback is sent whenever you change context on the deluge and whenever parameter values for the active context are changed.

Comes with an XML file with default CC to Deluge parameter mappings. You can customize this XML file to map CC's differently as required.

**Simple summary:** Set your channel(s), set your MIDI Controller(s) to the same channel(s), set a root note for your kits, confirm that your controller cc's are mapped to the parameters you want (via MIDIFollow.XML) and then play and control the deluge instruments and parameters with ease!

**No more re-learning your MIDI controllers every time you start a new song, add new clips or change instrument presets.**

- Note: Your existing MIDI learnings will remain untouched and can be used together with the master MIDI follow mode.

### Set Master MIDI Follow Channel(s), Set Kit Root Note, Enable/Disable Pop-ups and Enable/Disable MIDI Feedback

To use MIDI follow mode, you will need to configure the various MIDI Follow Mode settings by entering the settings menu and going to the sub menu for MIDI -> MIDI-Follow

- In the MIDI-Follow > Channel submenu, set the channel between 1 and 16 for Non-MPE or MPE Lower/Upper for MPE, for Synths (will apply to Synth Clips, MIDI Clips, and CV Clips), Kits and Params.
- In the MIDI-Follow > Kit Root Note submenu, set the root note for kits between 1 and 127 in order to map MIDI Notes received to Kit rows. The root note corresponds to the bottom row in a Kit.
- In the MIDI-Follow > Display Param submenu, enable or disable param pop-ups
- In the MIDI-Follow > Feedback > Feedback submenu, enable or disable MIDI follow feedback
- In the MIDI-Follow > Feedback > Automation Feedback submenu, enabled or disable MIDI follow feedback for automated parameters and set the rate at which feedback for automated parameters is sent
- In the MIDI-Follow > Feedback > Filter Responses submenu, enable or disable filtering of responses received within 1 second of sending a MIDI feedback value update.

### **Input Device Differentiation**
If you wish to use Input Device Differentiation with MIDI Follow Mode, you will need to learn your device to the required MIDI Follow Channel(s).

To learn the device, you need to enter the MIDI Follow Channel submenu, hold the Learn button and then send a Note or CC.

For example, if you enter the channel submenu for a Synth or Kit (e.g. `MIDI-Follow > Channel > Synth` or `MIDI-Follow > Channel > Kit`) and then `press and and hold the Learn button` and then send a Note to the deluge, the Deluge will update the the Synth channel to match the device and, if you are using an OLED Deluge, it will display the Device that has been learned on the screen.

### **Learning a Channel / Root Note**

As explained above, in the MIDI-Follow Channel submenu's you can hold learn to learn a device for input differentiation. You can also use this same functionality with input differentiation off to quickly update the MIDI-Follow channel's to match the channel used by your device. Simply hold Learn and send a Note or CC for any of the channels.

You can also use this method for updating the Kit Root Note. In the Kit Root Note menu, hold Learn and send a note from your device to update the Kit Root Note.

### **Notes:**
Notes received on the master MIDI channel will play the instrument in the active clip (e.g. a synth, MIDI clip, cv clip, or all kit rows).

- Note 1: You can play a synth, kit, MIDI or cv clip without entering the clip from arranger or song view. Simply press and hold the clip in arranger or song view to preview the clip (as you would to change the parameters of that clip with the gold encoders) and then send notes from your MIDI controller.

- Note 2: For Kit's, the bottom Kit row is mapped by default to the root note C1 (note # 36). All kit rows above are mapped to note's incrementally (e.g. 36, 37, 38, etc.). This kit root note # is configurable (from 0 to 127) through the Kit Root Note submenu.

- Note 3: MIDI Follow mode will always send notes to the active clip. This means that if you leave or unselect a clip, you can still send notes to that clip because in the Deluge, that clip is still recognized as the last active clip.

### **CC's:**
CC's received on the master MIDI channel that have been mapped to a parameter will change the value of that parameter in the active context (e.g. song, arranger, audio clip, instrument clip).

The parameters are controlled only in the current context.

- So if you are controlling filter, for example, while in song view it will only control the song’s filter. If you enter a specific synth clip, it will only control that synths filter. If you are in a kit clip it will either control the entire kit or a specific row in that clip (depending on whether you have affect entire enabled or not)
- In other words it checks what context you’re in and controls the parameters of that context.

Note: You can control the parameters of a synth or kit clip without entering the clip from arranger or song view. Simply press and hold the clip in arranger or song view to preview the clip (as you would to change the parameters of that clip with the gold encoders) and then send MIDI cc's from your MIDI controller to adjust the parameters.

#### Default MIDI CC Mappings
A default set of MIDI CC # to Deluge Parameter mappings has been created for MIDI Follow Mode. When you launch the Deluge after installing the firmware with MIDI Follow Mode, an XML file will be created to the root folder of the SD card titled "MIDIFollow.XML"

The default mappings have taken into account standard MIDI CC to parameter mappings and usages. It has also taken into account reserving of MIDI CC's for future Deluge functionality / feature implementations.

The default MIDI CC to parameter mappings, as mapped the to the parameter shortcuts on the Deluge grid are as follows:

image

> * See Appendix for detailed listing and description of default CC # to Parameter mappings.

Here are the MIDI CC #'s that have been reserved for other purposes:

image

#### Adjust MIDI CC Mappings
MIDI CC mappings for MIDI Follow Mode are saved to the root of your SD card in an XML file called MIDIFollow.XML

Within MIDIFollow.XML, all Parameters that can mapped to a MIDI CC are listed. The MIDI CC value is enclosed between a Parameter XML tag - e.g. `<lpfFrequency>74</lpfFrequency>` indicates that MIDI CC 74 is mapped to the LPF Frequency parameter. Conversely when a value of 255 is entered (e.g. `<hpfFrequency>255</hpfFrequency>`) it indicates that no MIDI CC value has been mapped to that parameter.

image
You can manually edit the MIDIFollow.XML to enter your MIDI CC mappings to each Parameter. 

The defaults from MIDIFollow.XML are loaded automatically when you start the Deluge so you can begin controlling the deluge with your MIDI controller right away. 

Note: A parameter can only be mapped to one MIDI CC. Conversely, a MIDI CC can be mapped to multiple parameters.

#### Display Parameter Names and Values on Screen

> Note: MIDI pop-up's can be disabled in the MIDI Follow Display Param submenu.

For mapped MIDI CC's, a pop up is shown on the display whenever MIDI CC's are received to indicate the name of the parameter that is being controlled by that MIDI CC and the current value being set for that parameter.

Screen Shot 2023-12-04 at 2 32 25 AM

Note: if the MIDI CC being received is for a Parameter that cannot be controlled in the current context (e.g. trying to control Attack while in Song View), the pop-up message will say "Can't Control: Parameter Name".

Screen Shot 2023-12-04 at 2 32 03 AM

## Re-cap of functionality

1. Added new MIDI submenu for MIDI-Follow where you can: 

- Set the MIDI Follow Channel and Device for Synth Notes (which also applies to MIDI and CV clips)
- Set the MIDI Follow Channel and Device for Kit Notes
- Set the MIDI Follow Channel and Device for Param CC's
- Set the MIDI Follow Root Note for Kits
- Enable or Disable MIDI Follow Param Pop-up's
- Enable or Disable MIDI Follow Feedback
- Enable or Disable MIDI Follow Feedback for Automated Parameters and set the MIDI Feedback Update Rate
- Enable or Disable MIDI Follow Feedback Filtering of MIDI CC responses received within 1 second of sending feedback
2. When MIDI Follow Mode is enabled, a MIDI Follow Channel has been set, and you have mapped your MIDI CC's, your external controller's Notes and MIDI CC's will be automatically directed to control the Notes of the Active Instrument (e.g. Synth, Kit, MIDI, CV) or the Parameters of the Active View (e.g. Song View, Arranger View, Audio Clip View, Instrument Clip View).
3. By default, the root note for kit's is C1 for the bottom kit row but this can be configured in the MIDI Follow menu.
4. Pop-up's are shown on the display for mapped MIDI CC's to show the name of the parameter being controlled and value being set for the parameter. This can be disabled in the menu.
5. MIDI feedback is sent for mapped CC's when the active context changes, change presets, or you change the value of a mapped parameter on the deluge (e.g. using mod encoders or select encoder if you're int he menu). MIDI feedback can be disabled in the menu.
6. MIDI feedback for automated parameters can also be sent and can be enabled or disabled in the MIDI feedback sub menu. When enabled, you choose between 3 speeds at which to send feedback for automated parameters: Low (500 ms), Medium (150 ms), High (40 ms). Sending automated parameter feedback can be taxing on the deluge MIDI output system, so depending on the amount of automation you do, you may want to adjust the speed (e.g. slow it down) to not affect the performance of the Deluge.
7. MIDI feedback can cause an undesirable result with certain applications when the application responds back to the Deluge after the Deluge has sent it an updated value (Loopy Pro and Drambo on iPad are known to do this). This can cause lag in the deluge and potential feedback loops. To handle this, a toggable filter was added which ignores messages received for the same ccNumber within 1 second of sending a MIDI feedback update. If the application receiving the MIDI feedback update does not send responses back to the Deluge, then this setting should be set to Disabled in the MIDI Feedback Filter Responses sub menu.

## Appendix - List of Deluge Parameters with Default Mapped CC's

image

### CODE REVIEW

File: src/definitions_cxx.hpp [hunk 1]
```diff
@@ -863,6 +879,7 @@ enum CCNumber {
 };
 constexpr int32_t kNumCCNumbersIncludingFake = 124;
 constexpr int32_t kNumRealCCNumbers = 120;
+constexpr int32_t kMaxCCNumber = 127;
```
Comment (m-m-adams) [inline src/definitions_cxx.hpp:882]: This should be 124 - all CCs above 120 are dedicated commands (like all notes off) and not controllers

File: src/deluge/gui/l10n/l10n.cpp [hunk 2]
```diff
@@ -5,6 +5,31 @@
 namespace deluge::l10n {
 Language const* chosenLanguage = nullptr;
 
+std::string_view getView(const deluge::l10n::Language& language, deluge::l10n::String string) {
```
Comment (m-m-adams) [inline src/deluge/gui/l10n/l10n.cpp:8]: why did you move these?

File: src/deluge/gui/menu_item/midi/follow/follow_channel_kit.h [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+			this->setValue(this->getValue() - NUM_CHANNELS);
+		}
+		else if (this->getValue() < 0) {
+			this->setValue(this->getValue() + NUM_CHANNELS);
+		}
+		Number::selectEncoderAction(offset);
+	}
+
+	void unlearnAction() {
+		midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::KIT)].device = NULL;
```
Comment (m-m-adams) [inline src/deluge/gui/menu_item/midi/follow/follow_channel_kit.h:119]: need to unlearn channel too

File: src/deluge/gui/menu_item/midi/follow/follow_channel_param.h [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+#pragma once
+#include "gui/menu_item/integer.h"
+#include "gui/ui/sound_editor.h"
+#include "io/midi/midi_engine.h"
+
+class MIDIDevice;
+
+namespace deluge::gui::menu_item::midi {
+
+class FollowChannelParam final : public Integer {
```
Comment (m-m-adams) [inline src/deluge/gui/menu_item/midi/follow/follow_channel_param.h:26]: you should abstract this into a class that can be initialized for param/kit/synth, as is there is a lot of code duplication

File: src/deluge/io/midi/midi_follow.cpp [hunk 5]
```diff
@@ -0,0 +1,700 @@
...
+	if (!clip && useActiveClip) {
+		clip = currentSong->currentClip;
+	}
+	return clip;
+}
+
+/// based on the current context, as determined by clip returned from the getClipForMidiFollow function
+/// obtain the modelStackWithParam for that context and return it so it can be used by midi follow
+ModelStackWithAutoParam*
+MidiFollow::getModelStackWithParam(ModelStackWithThreeMainThings* modelStackWithThreeMainThings,
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_follow.cpp:133]: this would be much easier to follow if it was split into seperate functions

1. One requiring a clip and timeline counter, and one with a modelstack and no clip
2. When there is a clip, a seperate function to check params for each of audio/kit/synth/midi

File: src/deluge/io/midi/midi_follow.cpp [hunk 6]
```diff
@@ -0,0 +1,700 @@
...
+		Clip* clip = getClipForMidiFollow(true);
+		if (!on && !clip) {
+			// for note off's when you're no longer in an active clip,
+			// see if a note on message was previously sent so you can
+			// direct the note off to the right place
+			clip = clipForLastNoteReceived;
+		}
+
+		// Only send if not muted - but let note-offs through always, for safety
+		if (clip && (!on || currentSong->isOutputActiveInArrangement(clip->output))) {
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_follow.cpp:292]: I think it makes more sense to invert this logic - check match, if there is a match on kit or synth channel then get the clip and send it through. That will be faster to return in the common case that it's not a match anyway

File: src/deluge/io/midi/midi_follow.cpp [hunk 7]
```diff
@@ -0,0 +1,700 @@
...
+/// called from playback handler
+/// determines whether a midi cc received is midi follow relevant
+/// and should be routed to the active context for further processing
+void MidiFollow::midiCCReceived(MIDIDevice* fromDevice, uint8_t channel, uint8_t ccNumber, uint8_t value,
+                                bool* doingMidiThru, bool isMPE, ModelStack* modelStack) {
+	// midi follow mode
+	if (midiEngine.midiFollow) {
+		//obtain clip for active context (for params that's only for the active mod controllable stack)
+		Clip* clip = getClipForMidiFollow();
+		if (!isMPE && view.activeModControllableModelStack.modControllable) {
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_follow.cpp:352]: same here, check matches first

File: src/deluge/model/mod_controllable/mod_controllable_audio.cpp [hunk 8]
```diff
@@ -1681,137 +1681,368 @@ bool ModControllableAudio::offerReceivedCCToLearnedParams(MIDIDevice* fromDevice
...
+	return messageUsed;
+}
 
-						Similarly a Midi CC Value of 127 is equal to a Deluge Knob Position Value of +63 (127 less 64)
+/// this function is called when midi follow is enabled
+/// it checks if the ccNumber received has been learned to any parameters in midi learning view
+/// if the cc has been learned, it sets the new value for that parameter
+/// this function works by first checking the active context to see if there is an active clip
+/// to determine if the cc intends to control a song level or clip level parameter
+void ModControllableAudio::offerReceivedCCToMidiFollow(ModelStack* modelStack, Clip* clip, int32_t ccNumber,
```
Comment (m-m-adams) [inline src/deluge/model/mod_controllable/mod_controllable_audio.cpp:1730]: This is no longer required right? Since now it's being called from the playback handler

File: src/definitions_cxx.hpp (hunk 1 above)
Comment (seangoodvibes) [reply src/definitions_cxx.hpp:882]: I've since changed this to a more general midi value range constant kMaxMIDIValue = 127;

If I used a smaller number for Midi CC's as you suggest, is there a possibility someone could try to send a CC above 124?

I use this constant when defining a couple array's to track some info by CC #.

	int32_t previousKnobPos[kMaxMIDIValue + 1];
	uint32_t timeLastCCSent[kMaxMIDIValue + 1];

File: src/deluge/gui/l10n/l10n.cpp (hunk 2 above)
Comment (seangoodvibes) [reply src/deluge/gui/l10n/l10n.cpp:8]: @PaulFreund provided me with an l10n patch to add to my PR. It was to reduce the binary size impact of the string localization

File: src/deluge/gui/menu_item/midi/follow/follow_channel_kit.h (hunk 3 above)
Comment (seangoodvibes) [reply src/deluge/gui/menu_item/midi/follow/follow_channel_kit.h:126]: Done!

File: src/deluge/model/mod_controllable/mod_controllable_audio.cpp (hunk 8 above)
Comment (seangoodvibes) [reply src/deluge/model/mod_controllable/mod_controllable_audio.cpp]: as discussed, the function is required and to avoid confusion I've now renamed the function to "receivedCCFromMidiFollow" to accurately describe what the function is doing

File: src/deluge/io/midi/midi_follow.cpp (hunk 6 above)
Comment (seangoodvibes) [reply src/deluge/io/midi/midi_follow.cpp]: Done

File: src/deluge/io/midi/midi_follow.cpp (hunk 5 above)
Comment (seangoodvibes) [reply src/deluge/io/midi/midi_follow.cpp:131]: done!

File: src/deluge/gui/menu_item/midi/follow/follow_channel_param.h (hunk 4 above)
Comment (seangoodvibes) [reply src/deluge/gui/menu_item/midi/follow/follow_channel_param.h]: done!

File: src/definitions_cxx.hpp (hunk 1 above)
Comment (m-m-adams) [reply src/definitions_cxx.hpp]: Generally deluge should never send a CC above 120 except as part of a separate command that doesn't need tracking, I said 124 as there are dedicated "CCs" internally for MPE params 

Leaving it as 127 in array definitions is fine

File: src/deluge/storage/flash_storage.cpp [hunk 9]
```diff
@@ -170,6 +182,19 @@ void resetSettings() {
 	PadLEDs::flashCursor = FLASH_CURSOR_SLOW;
 
 	midiEngine.midiThru = false;
+	midiEngine.midiFollow = false;
+	for (auto& midiChannelType : midiEngine.midiFollowChannelType) {
+		midiChannelType.clear();
+	}
+	midiEngine.midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::SYNTH)].channelOrZone = 15;
```
Comment (m-m-adams) [inline src/deluge/storage/flash_storage.cpp:189]: init these to 255 (unlearnt)

File: src/deluge/io/midi/midi_engine.cpp [hunk 10]
```diff
@@ -218,6 +218,18 @@ MidiEngine::MidiEngine() {
 	lastStatusByteSent = 0;
 	currentlyReceivingSysExSerial = false;
 	midiThru = false;
+	midiFollow = false;
+	for (auto& midiChannelType : midiEngine.midiFollowChannelType) {
+		midiChannelType.clear();
+	}
+	midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::SYNTH)].channelOrZone = 15;
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_engine.cpp:225]: init to un learnt instead

File: src/deluge/gui/ui/menus.cpp [hunk 11]
```diff
@@ -755,6 +759,58 @@ ToggleBool midiThruMenu{STRING_FOR_MIDI_THRU, STRING_FOR_MIDI_THRU, midiEngine.m
 // MIDI Takeover
 midi::Takeover midiTakeoverMenu{STRING_FOR_TAKEOVER};
 
+//MIDI Follow
+ToggleBool midiFollowMenu{STRING_FOR_FOLLOW, STRING_FOR_FOLLOW, midiEngine.midiFollow};
```
Comment (m-m-adams) [inline src/deluge/gui/ui/menus.cpp:763]: Not needed, it's disabled if the channel isn't set

File: src/deluge/io/midi/midi_engine.cpp [hunk 12]
```diff
@@ -218,13 +218,9 @@ MidiEngine::MidiEngine() {
 	lastStatusByteSent = 0;
 	currentlyReceivingSysExSerial = false;
 	midiThru = false;
-	midiFollow = false;
 	for (auto& midiChannelType : midiEngine.midiFollowChannelType) {
 		midiChannelType.clear();
 	}
-	midiFollowChannelType[util::to_underlying(MIDIFollowChannelType::SYNTH)].channelOrZone = 15;
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_engine.cpp]: these should be initialized to none (255)

Comment (seangoodvibes) [reply src/deluge/io/midi/midi_engine.cpp]: for (auto& midiChannelType : midiEngine.midiFollowChannelType) {
		midiChannelType.clear();
	}
	
	this code takes care of that :)

File: src/deluge/io/midi/midi_follow.cpp [hunk 13]
```diff
@@ -347,43 +347,40 @@ PLACE_SDRAM_DATA Clip* clipForLastNoteReceived[kMaxMIDIValue + 1] = {0};
...
-					}
-					if (on) {
-						clipForLastNoteReceived[note] = clip;
-					}
+		// Only send if not muted - but let note-offs through always, for safety
+		if (clip && (clip->output->type != InstrumentType::AUDIO)
+		    && (!on || currentSong->isOutputActiveInArrangement(clip->output))) {
+			ModelStackWithTimelineCounter* modelStackWithTimelineCounter = modelStack->addTimelineCounter(clip);
+			//Output is a kit or melodic instrument
+			//Note: Midi instruments not currently supported for midi follow mode
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_follow.cpp:368]: incorrect comment - midi instruments are just melodic instruments

Comment (seangoodvibes) [reply src/deluge/io/midi/midi_follow.cpp:368]: oops old comment, will remove now thx

File: src/deluge/io/midi/midi_follow.cpp [hunk 14]
```diff
@@ -347,43 +347,40 @@ PLACE_SDRAM_DATA Clip* clipForLastNoteReceived[kMaxMIDIValue + 1] = {0};
...
-			if (on) {
-				clip = getClipForMidiFollow(true);
-			}
-			else {
-				// for note off's, see if a note on message was previously sent so you can
-				// direct the note off to the right place
-				clip = clipForLastNoteReceived[note];
-			}
+	MIDIMatchType matchSynth = checkMidiFollowMatch(fromDevice, channel, MIDIFollowChannelType::SYNTH);
+	MIDIMatchType matchKit = checkMidiFollowMatch(fromDevice, channel, MIDIFollowChannelType::KIT);
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_follow.cpp:351]: I think the logic in here can be improved - most of the time both channels will be the same, and when they're not just sending the note to midi follow in both cases makes more sense than dropping it. I'd suggest updating your check match function to just return a single match type

Alternatively if you tracked the last kit and last synth separately this method would make sense

File: src/deluge/io/midi/midi_follow.cpp [hunk 15]
```diff
@@ -414,50 +411,46 @@ void MidiFollow::offerReceivedNoteToMelodicInstrument(ModelStackWithTimelineCoun
...
-		MIDIMatchType matchParam = checkMidiFollowMatch(fromDevice, channel, MIDIFollowChannelType::PARAM);
+	MIDIMatchType matchSynth = checkMidiFollowMatch(fromDevice, channel, MIDIFollowChannelType::SYNTH);
+	MIDIMatchType matchKit = checkMidiFollowMatch(fromDevice, channel, MIDIFollowChannelType::KIT);
+	MIDIMatchType matchParam = checkMidiFollowMatch(fromDevice, channel, MIDIFollowChannelType::PARAM);
+	if ((matchSynth != MIDIMatchType::NO_MATCH)
+	    || (matchKit == MIDIMatchType::MPE_MASTER || matchKit == MIDIMatchType::MPE_MEMBER)
+	    || (matchParam != MIDIMatchType::NO_MATCH)) {
+		//obtain clip for active context (for params that's only for the active mod controllable stack)
+		Clip* clip = getClipForMidiFollow();
+		if (!isMPE && view.activeModControllableModelStack.modControllable && (matchParam != MIDIMatchType::NO_MATCH)) {
```
Comment (m-m-adams) [inline src/deluge/io/midi/midi_follow.cpp:421]: matchParam should have to be master or channel

**Review by m-m-adams — APPROVED** [review]

**Comment by Queroland** [comment]
Hey...Happy new years and OMG! today is christmas again :) It`s sooo much fun and just ..wow! I am probably the first after you active-programming guys who tried it :) Used your build from here as its not in the nightly.  I used the midi fighter twister and also the led feedback is really adapting to what I press..it works like charm. Such a big day today :) Also in kit affect entire or row, updating, can record automations and the leds are flickering :) Thank you very much sir!
Why did you decide against the learning view - you already had it done.
One last thing I wanted to know: why did you close the snake game weeks before?
Would you participate from the donations? :)
Greetings, Roland from germany

**Comment by seangoodvibes** [comment]
> Hey...Happy new years and OMG! today is christmas again :) It`s sooo much fun and just ..wow! I am probably the first after you active-programming guys who tried it :) Used your build from here as its not in the nightly. I used the midi fighter twister and also the led feedback is really adapting to what I press..it works like charm. Such a big day today :) Also in kit affect entire or row, updating, can record automations and the leds are flickering :) Thank you very much sir! Why did you decide against the learning view - you already had it done. One last thing I wanted to know: why did you close the snake game weeks before? Would you participate from the donations? :) Greetings, Roland from germany

Hey! Great to hear that its s working for you :D! Makes me want to get a midi fighter twister ahah.

As for removing learning view - I know it was done and I really liked it, but we're trying to keep the deluge performance as high as possible which means reducing the size impact of new features. The learning view wasn't regarded as a "must have" in this case and by removing it, it simplified the feature's code quite a bit. I know it was definitely useful for learning your controllers, but alas it is gone :).

As for the snake game, I just closed it for now because I don't have the time to work on it. I focused my time on performance view, this feature, automation view for audio clips and other things improvements :). 

I'm sure I'll get back to snake game one of these days ahah.
